libs/partners/azure-dynamic-sessions/langchain_azure_dynamic_sessions/tools/sessions.py

```diff
@@ -0,0 +1,256 @@
+    def full_path(self) -> str:
+        """Get the full path of the file."""
+        return f"/mnt/data/{self.filename}"
+
+    @staticmethod
+    def from_dict(data: dict) -> "RemoteFileMetadata":
+        """Create a RemoteFileMetadata object from a dictionary."""
+        properties = data.get("properties", {})
+        return RemoteFileMetadata(
+            filename=properties.get("filename"),
+            size_in_bytes=properties.get("size"),
+        )
+
+
+class SessionsPythonREPLTool(BaseTool):
+    """A tool for running Python code in an Azure Container Apps dynamic sessions
+    code interpreter."""
+
+    name: str = "Sessions_Python_REPL"
+    description: str = (
+        "A Python shell. Use this to execute python commands "
+        "when you need to perform calculations or computations. "
+        "Input should be a valid python command. "
+        "Returns the result, stdout, and stderr. "
+    )
+
+    sanitize_input: bool = True
+    """Whether to sanitize input to the python REPL."""
+
+    pool_management_endpoint: str
+    """The management endpoint of the session pool. Should end with a '/'."""
+
+    access_token_provider: Callable[
+        [], Optional[str]
+    ] = _access_token_provider_factory()
+    """A function that returns the access token to use for the session pool."""
+
+    session_id: str = str(uuid4())
+    """The session ID to use for the session pool. Defaults to a random UUID."""
+
+    def _build_url(self, path) -> str:
+        pool_management_endpoint = self.pool_management_endpoint
+        if not pool_management_endpoint:
+            raise ValueError("pool_management_endpoint is not set")
+        if not pool_management_endpoint.endswith("/"):
+            pool_management_endpoint += "/"
+        encoded_session_id = urllib.parse.quote(self.session_id)
+        query = f"identifier={encoded_session_id}&api-version=2024-02-02-preview"
+        query_separator = "&" if "?" in pool_management_endpoint else "?"
+        full_url = pool_management_endpoint + path + query_separator + query
+        return full_url
+
+    def _run(self, python_code: str) -> Any:
+        if self.sanitize_input:
+            python_code = _sanitize_input(python_code)
+
+        access_token = self.access_token_provider()
+        api_url = self._build_url("code/execute")
+        headers = {
+            "Authorization": f"Bearer {access_token}",
+            "Content-Type": "application/json",
+            "User-Agent": USER_AGENT,
+        }
+        body = {
+            "properties": {
+                "codeInputType": "inline",
+                "executionType": "synchronous",
+                "code": python_code,
+            }
+        }
+
+        response = requests.post(api_url, headers=headers, json=body)
+        response.raise_for_status()
+        response_json = response.json()
+        properties = response_json.get("properties", {})
+        return (
+            f"result:\n{properties['result']}\n\n"
```

Comment:
should we filter out images or anything else that doesn't really make sense to have in str?

Reply:
also maybe simpler/better to just `json.dumps(..., indent=2)` the dict instead of custom formatting it? most models do pretty well with reading json formatted outputs